File: Algo_DBMS/Algo_Lab_Sheet_10/lab_Sheet_One.java

import java.util.*;

public class lab_Sheet_One
{

    // Bubble Sort
    public static int bubbleSort (int [] array)
    {
        int k = array.length;
        int stepCount = 0;
        for (int i = 0; i < k; i++)
        {
            boolean swapped = false;
            for (int j = 0; j < (k - i - 1); j++)
            {
                stepCount++; // Comparison
                if (array [j] > array [j + 1])
                {
                    int temp = array [j];
                    array [j] = array [j + 1];
                    array [j + 1] = temp;
                    stepCount++; // Swap
                    swapped = true;
                }
            }
            if (!swapped) break;
        }
        return stepCount;
    }

    // Selection Sort
    public static int selectionSort (int [] array)
    {
        int k = array.length;
        int stepCount = 0;
        for (int i = 0; i < k; i++)
        {
            int minIdx = i;
            for (int j = (i + 1); j < k; j++)
            {
                stepCount++; // Comparison
                if (array [j] < array [minIdx])
                {
                    minIdx = j;
                }
            }
            if (minIdx != i)
            {
                int temp = array[i];
                array [i] = array [minIdx];
                array [minIdx] = temp;
                stepCount++; // Swap
            }
        }
        return stepCount;
    }

    // Insertion Sort
    public static int insertionSort (int [] array)
    {
        int k = array.length;
        int stepCount = 0;
        for (int i = 1; i < k; i++)
        {
            int key = array [i];
            int j = (i - 1);
            while ((j >= 0) && (array [j] > key))
            {
                stepCount++; // Comparison
                array [j + 1] = array [j];
                stepCount++; // Assignment
                j--;
            }
            array [j + 1] = key;
            stepCount++; // Assignment
        }
        return stepCount;
    }

    // Function to measure step counts for different input sizes
    public static void measureStepCounts ()
    {
        int [] inputSizes = {10, 20, 50};
        String [] algorithms = {"Bubble Sort", "Selection Sort", "Insertion Sort"};
        String [] cases = {"Best", "Worst", "Average"};

        // Results storage
        List <Integer> [] bubbleResults = new ArrayList [3];
        List <Integer> [] selectionResults = new ArrayList [3];
        List <Integer> [] insertionResults = new ArrayList [3];

        for (int i = 0; i < 3; i++)
        {
            bubbleResults [i] = new ArrayList <> ();
            selectionResults [i] = new ArrayList <> ();
            insertionResults [i] = new ArrayList <> ();
        }

        for (int size : inputSizes)
        {
            // Generate test cases
            int [] bestCase = new int [size];
            int [] worstCase = new int [size];
            int [] averageCase = new int [size];

            // Best case: Already sorted
            for (int i = 0; i < size; i++) bestCase [i] = (i + 1);

            // Worst case: Reverse sorted
            for (int i = 0; i < size; i++) worstCase [i] = (size - i);

            // Average case: Randomized
            Random random = new Random ();
            for (int i = 0; i < size; i++) averageCase [i] = random.nextInt(100) + 1;

            // Bubble Sort
            bubbleResults [0].add (bubbleSort (Arrays.copyOf (bestCase, bestCase.length)));
            bubbleResults [1].add (bubbleSort (Arrays.copyOf (worstCase, worstCase.length)));
            bubbleResults [2].add (bubbleSort (Arrays.copyOf (averageCase, averageCase.length)));

            // Selection Sort
            selectionResults [0].add (selectionSort (Arrays.copyOf (bestCase, bestCase.length)));
            selectionResults [1].add (selectionSort (Arrays.copyOf (worstCase, worstCase.length)));
            selectionResults [2].add (selectionSort (Arrays.copyOf (averageCase, averageCase.length)));

            // Insertion Sort
            insertionResults [0].add (insertionSort (Arrays.copyOf (bestCase, bestCase.length)));
            insertionResults [1].add (insertionSort (Arrays.copyOf (worstCase, worstCase.length)));
            insertionResults [2].add (insertionSort (Arrays.copyOf (averageCase, averageCase.length)));
        }

        // Print results
        System.out.println ("Bubble Sort : ");
        for (int i = 0; i < 3; i++)
        {
            System.out.print (cases [i] + " :  ");
            for (int count : bubbleResults [i]) System.out.print (count + " ");
            System.out.println ();
        }

        System.out.println ("\nSelection Sort : ");
        for (int i = 0; i < 3; i++)
        {
            System.out.print (cases [i] + " :  ");
            for (int count : selectionResults [i]) System.out.print (count + " ");
            System.out.println ();
        }

        System.out.println ("\nInsertion Sort : ");
        for (int i = 0; i < 3; i++)
        {
            System.out.print (cases [i] + " :  ");
            for (int count : insertionResults [i]) System.out.print (count + " ");
            System.out.println ();
        }
    }

    public static void main(String[] args)
    {
        measureStepCounts();
    }
}

File: Algo_DBMS/Algo_Lab_Sheet_10/lab_Sheet_One.cpp

# include <iostream>
# include <vector>
# include <algorithm>
# include <cstdlib>
# include <ctime>
using namespace std;

// Bubble Sort
int bubble_sort (vector <int> & array)
{
    int k = array.size();
    int step_count = 0;
    for (int i = 0; i < k; ++i)
    {
        bool swapped = false;
        for (int j = 0; j < (k - i - 1); ++j)
        {
            step_count++; // Comparison
            if (array [j] > array [j + 1])
            {
                swap(array [j], array [j + 1]);
                step_count++; // Swap
                swapped = true;
            }
        }
        if (!swapped) break;
    }
    return step_count;
}

// Selection Sort
int selection_sort (vector <int> & array)
{
    int k = array.size();
    int step_count = 0;
    for (int i = 0; i < k; ++i)
    {
        int min_idx = i;
        for (int j = (i + 1); j < k; ++j)
        {
            step_count++; // Comparison
            if (array [j] < array [min_idx])
            {
                min_idx = j;
            }
        }
        if (min_idx != i)
        {
            swap(array [i], array [min_idx]);
            step_count++; // Swap
        }
    }
    return step_count;
}

// Insertion Sort
int insertion_sort (vector <int> & array)
{
    int k = array.size();
    int step_count = 0;
    for (int i = 1; i < k; ++i)
    {
        int key = array[i];
        int j = (i - 1);
        while ((j >= 0) && (array [j] > key))
        {
            step_count++; // Comparison
            array [j + 1] = array [j];
            step_count++; // Assignment
            --j;
        }
        array [j + 1] = key;
        step_count++; // Assignment
    }
    return step_count;
}

// Function to measure step counts for different input sizes
void measure_step_counts ()
{
    vector <int> input_sizes = {10, 20, 50};
    vector <string> algorithms = {"Bubble Sort", "Selection Sort", "Insertion Sort"};
    vector <string> cases = {"Best", "Worst", "Average"};

    // Results storage
    vector <vector <int> > bubble_results (3, vector <int> ());
    vector <vector <int> > selection_results (3, vector <int> ());
    vector <vector <int> > insertion_results (3, vector <int> ());

    for (int size : input_sizes)
    {
        // Generate test cases
        vector <int> best_case (size);
        vector <int> worst_case (size);
        vector <int> average_case (size);
        // Best case: Already sorted
        for (int i = 0; i < size; ++i) best_case [i] = (i + 1);
        // Worst case: Reverse sorted
        for (int i = 0; i < size; ++i) worst_case [i] = (size - i);
        // Average case: Randomized
        srand (time (0));
        for (int i = 0; i < size; ++i) average_case [i] = rand() % 100 + 1;

        // Bubble Sort
        bubble_results [0].push_back (bubble_sort (best_case));
        bubble_results [1].push_back (bubble_sort (worst_case));
        bubble_results [2].push_back (bubble_sort (average_case));

        // Selection Sort
        selection_results [0].push_back (selection_sort (best_case));
        selection_results [1].push_back (selection_sort (worst_case));
        selection_results [2].push_back (selection_sort (average_case));

        // Insertion Sort
        insertion_results [0].push_back (insertion_sort (best_case));
        insertion_results [1].push_back (insertion_sort (worst_case));
        insertion_results [2].push_back (insertion_sort (average_case));
    }

    // Print results
    cout << "Bubble Sort : \n";
    for (int i = 0; i < 3; ++i)
    {
        cout << cases [i] << " :  ";
        for (int count : bubble_results[i]) cout << count << " ";
        cout << endl;
    }

    cout << "\nSelection Sort : \n";
    for (int i = 0; i < 3; ++i)
    {
        cout << cases [i] << " :  ";
        for (int count : selection_results[i]) cout << count << " ";
        cout << endl;
    }

    cout << "\nInsertion Sort : \n";
    for (int i = 0; i < 3; ++i)
    {
        cout << cases [i] << " :  ";
        for (int count : insertion_results[i]) cout << count << " ";
        cout << endl;
    }
}

int main()
{
    measure_step_counts();
    return 0;
}
